### DISCUSSION TITLE
Question about websocket read/writes thread safety.

### DISCUSSION THREAD

**Discussion by kujukuju** [cesanta/mongoose#1942] [Ask questions, get answers] [answered]
HI, I'm trying to figure out if it's thread safe to call mg_ws_write from a completely different thread than the thread doing the reads.

I read through a bunch of issues/questions here and my guess so far is that this is fine. It seems that the websocket send events fire on the poll event. Presumably, as long as I'm mg_ws_write ing from the same thread, I don't think it could have any async issues.

However, then I saw this comment:
https://github.com/cesanta/mongoose/pull/139#issuecomment-20871332

1. Would be any issues calling mg_ws_write without dealing with thread safety?

2. I have a loop set up calling mg_poll with a 0 ms timeout. All packet reads into here get put into a circular buffer than I then access from my other thread safely.

If a read takes say, 2 milliseconds, I believe this would delay the actual tcp packet sending from my mg_ws_write by 2 milliseconds due to packet writes waiting on the poll event to be processed. I saw another issue here about waking up the manager in order to perform an immediate write but the referenced wakeup function no longer seems to exist.

Is there a way to wake up the manager to immediately send a ws packet without waiting on the next poll?

Thanks!

**Comment by scaprile** [comment, marked as answer]
From [the docs](https://mongoose.ws/documentation/#2-minute-integration-guide)
> NOTE: Since Mongoose's core is not protected against concurrent accesses, make sure that all mg_* API functions are called from the same thread or RTOS task.

If you absolutely need multi-threading, take the time to go through the user guide (linked above) and [tutorials](https://mongoose.ws/tutorials/), there is one that shows you how to do [multi-threading](https://mongoose.ws/tutorials/multi-threaded/).

Please read the official docs instead of comments/issues from 2013, 10 years ago. Mongoose is 7.8 now

Your OS provides a socket environment and a select()/poll() call. Mongoose will wait on those sockets, so reads are not delayed.
Writes are immediately sent to the socket if UDP, and to a buffer if TCP. The poll timeout is the time mg_mgr_poll() provides to select()/poll(), that is, the time it will wait before returning when there is no pending action (read or write) on the active sockets
For a WS socket (TCP), the write will go to a buffer and when there is room for data to send, the socket will signal it is ready for writing and Mongoose will take your data from the buffer and write it to the socket.

The multi-threading example uses a pipe made from wrapping 2 UDP sockets connected back to back. You send() to one end and Mongoose waits on the other end.

**Reply by kujukuju** [reply]
Thanks @scaprile, I believe that answers my question.
